From 1abe90b7275b8771f5e5c588447f85a5867b2ce7 Mon Sep 17 00:00:00 2001
From: Jaroslav Kysela <jkysela@redhat.com>
Date: Wed, 3 Sep 2014 07:28:23 -0400
Subject: [alsa] hda/realtek: Fix the noise after suspend and resume on ALC282 codec

Message-id: <140972823573-239-patchset-jkysela@redhat.com>
Patchwork-id: 90939
O-Subject: [RHEL 7.1 PATCH 238/442] ALSA: hda/realtek - Fix the noise after suspend and resume on ALC282 codec
Bugzilla: 1112200
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>

When the power state of ALC283 codec goes to D3 or return back to D0,
it gives a noise via headphone output. To follow the depop procedure,
it will be better.

Signed-off-by: Kailang Yang <kailang@realtek.com>
Signed-off-by: Takashi Iwai <tiwai@suse.de>

(picked from kernel: 7b5c7a0240b11b382073361c4ba9257c42d057e9)
rhbz#1112200
Signed-off-by: Jaroslav Kysela <jkysela@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/sound/pci/hda/patch_realtek.c b/sound/pci/hda/patch_realtek.c
index b08bdcd..be663cd 100644
--- a/sound/pci/hda/patch_realtek.c
+++ b/sound/pci/hda/patch_realtek.c
@@ -2787,6 +2787,77 @@ static void alc269_shutup(struct hda_codec *codec)
  snd_hda_shutup_pins(codec);
 }
 
+static void alc282_init(struct hda_codec *codec)
+{
+ struct alc_spec *spec = codec->spec;
+ hda_nid_t hp_pin = spec->gen.autocfg.hp_pins[0];
+ bool hp_pin_sense;
+ int coef78;
+
+ if (!hp_pin)
+  return;
+ hp_pin_sense = snd_hda_jack_detect(codec, hp_pin);
+ coef78 = alc_read_coef_idx(codec, 0x78);
+
+ /* Index 0x78 Direct Drive HP AMP LPM Control 1 */
+ /* Headphone capless set to high power mode */
+ alc_write_coef_idx(codec, 0x78, 0x9004);
+
+ if (hp_pin_sense)
+  msleep(2);
+
+ snd_hda_codec_write(codec, hp_pin, 0,
+       AC_VERB_SET_AMP_GAIN_MUTE, AMP_OUT_MUTE);
+
+ if (hp_pin_sense)
+  msleep(85);
+
+ snd_hda_codec_write(codec, hp_pin, 0,
+       AC_VERB_SET_PIN_WIDGET_CONTROL, PIN_OUT);
+
+ if (hp_pin_sense)
+  msleep(100);
+
+ /* Headphone capless set to normal mode */
+ alc_write_coef_idx(codec, 0x78, coef78);
+}
+
+static void alc282_shutup(struct hda_codec *codec)
+{
+ struct alc_spec *spec = codec->spec;
+ hda_nid_t hp_pin = spec->gen.autocfg.hp_pins[0];
+ bool hp_pin_sense;
+ int coef78;
+
+ if (!hp_pin) {
+  alc269_shutup(codec);
+  return;
+ }
+
+ hp_pin_sense = snd_hda_jack_detect(codec, hp_pin);
+ coef78 = alc_read_coef_idx(codec, 0x78);
+ alc_write_coef_idx(codec, 0x78, 0x9004);
+
+ if (hp_pin_sense)
+  msleep(2);
+
+ snd_hda_codec_write(codec, hp_pin, 0,
+       AC_VERB_SET_AMP_GAIN_MUTE, AMP_OUT_MUTE);
+
+ if (hp_pin_sense)
+  msleep(85);
+
+ snd_hda_codec_write(codec, hp_pin, 0,
+       AC_VERB_SET_PIN_WIDGET_CONTROL, 0x0);
+
+ if (hp_pin_sense)
+  msleep(100);
+
+ alc_auto_setup_eapd(codec, false);
+ snd_hda_shutup_pins(codec);
+ alc_write_coef_idx(codec, 0x78, coef78);
+}
+
 static void alc283_restore_default_value(struct hda_codec *codec)
 {
  int val;
@@ -4724,6 +4795,8 @@ static int patch_alc269(struct hda_codec *codec)
   break;
  case 0x10ec0282:
   spec->codec_variant = ALC269_TYPE_ALC282;
+  spec->shutup = alc282_shutup;
+  spec->init_hook = alc282_init;
   break;
  case 0x10ec0233:
  case 0x10ec0283:
-- 
1.7.1